### PR TITLE
🗒️ flesh out the documentation for quilt_rails

### PR DESCRIPTION
**Pull request by TheMallen**
This PR reorganizes and fleshes out the documentation for `quilt_rails`, adding some guidelines for testing, a table of contents, and stronger guidelines for how to layout a complex application.

### CODE REVIEW

File: gems/quilt_rails/README.md [hunk 1]
```diff
@@ -119,21 +142,204 @@ function App() {
...
+#### `sewing_kit:install`
 
 Adds a basic `sewing-kit.config.ts` file.
 
 ## Advanced use
 
-### Interacting with the request / response in your React code
+### Testing
+
+For fast tests with consistent results, test front-end components using the tools provided by sewing-kit instead of Rails integration tests.
```
Comment (TheMallen) [inline gems/quilt_rails/README.md]: some of this stuff could link out to `sewing-kit`'s docs but I think it's more likely to be read if it is duplicated here

File: gems/quilt_rails/README.md [hunk 2]
```diff
@@ -186,9 +392,16 @@ function App() {
 export default App;
 ```
 
+### Isomorphic state
+
+With SSR enabled React apps, state must be serialized on the server and deserialized on the client to keep it consistent. When using `@shopify/react-server`, the best tool for this job is [`@shopify/react-html`](https://github.com/Shopify/quilt/tree/master/packages/react-html)'s [`useSerialized`](https://github.com/Shopify/quilt/tree/master/packages/react-html#in-your-application-code) hook.
```
Comment (TheMallen) [inline gems/quilt_rails/README.md]: I think this section is still pretty impossible to understand as an outsider so we probably want to add some real examples here. We may want to have a `recipes` section or something like that in the future.

File: gems/quilt_rails/README.md [hunk 3]
```diff
@@ -186,9 +392,16 @@ function App() {
...
+
+
 ### Customizing the node server
 
-By default, sewing-kit bundles in `@shopify/react-server-webpack-plugin` for `quilt_rails` applications to get you up and running fast without needing to manually write any node server code. If you would like to customize what data your react application receives from the incoming request, you can add your own `server.js` / `server.ts` file to the app folder.
+By default, sewing-kit bundles in `@shopify/react-server-webpack-plugin` for `quilt_rails` applications to get apps up and running fast without needing to manually write any node server code. If what it provides is not sufficient, a custom server can be defined by adding a `server.js` or `server.ts` file to the app folder.
```
Comment (TheMallen) [inline gems/quilt_rails/README.md]: not sure if adding a `customizing the client entrypoint` would be useful too, I don't think we've ever needed to and most things you could do there you could also do in an if statement in the index file /shrug

File: gems/quilt_rails/README.md [hunk 4]
```diff
@@ -2,47 +2,79 @@
...
-# Optional - add Polaris and quilt libraries
-yarn add @shopify/polaris @shopify/react-self-serializers react react-dom
+# Add Polaris and React
+yarn add @shopify/polaris react react-dom
 
 yarn
 dev up
 ```
 
-### Add JavaScript
```
Comment (TheMallen) [inline gems/quilt_rails/README.md]: I'm not sure why we had this here, it was pretty out of place

File: gems/quilt_rails/README.md [hunk 5]
```diff
@@ -78,27 +110,18 @@ class ReactController < ApplicationController
 end
 ```
 
-Have your routes wired up to default to your react controller.
+Add routes to default to the `ReactController`.
```
Comment (cartogram) [inline gems/quilt_rails/README.md]: Not crazy about the wording here, maybe 

```suggestion
Add routes to point to your `ReactController` by default
```

File: gems/quilt_rails/README.md [hunk 6]
```diff
@@ -78,27 +110,18 @@ class ReactController < ApplicationController
...
-├── Gemfile (must contain "gem 'sewing_kit" and "gem 'quilt_rails'")
-├── package.json (must specify '@shopify/sewing-kit' and `@shopify/react-server` as 'dependencies')
-│
-└── app
-   └── ui
-   │   └─- index.js
-   └── controllers
-       └─- react_controller (see above)
-```
+`sewing_kit` looks for the top level component of your React app in `app/ui/index`. The component exported from this component (and any imported JS/CSS) will be built into a `main` bundle, and used to render the initial server-rendered markup.
```
Comment (cartogram) [inline gems/quilt_rails/README.md]:
```suggestion
`sewing_kit` looks for the top level component of your React app in `app/ui/index`. The component exported from this file (and any imported JS/CSS) will be built into a `main` bundle, and used to render the initial server-rendered markup.
```

File: gems/quilt_rails/README.md [hunk 7]
```diff
@@ -119,21 +142,204 @@ function App() {
...
+        ├─- index.js (exports a React component)
+        ├── styles (optional)
+        │   └── settings.scss (global vars and @polaris overrides)
+        │
+        └── tests (optional)
+        │   └── each-test.ts
+        │   └── setup.ts
+        └── features (optional)
+            ├── App
+            │   ├── index.js
```
Comment (cartogram) [inline gems/quilt_rails/README.md]: should we do like `index.{js|ts}`?

File: gems/quilt_rails/README.md [hunk 8]
```diff
@@ -119,21 +142,204 @@ function App() {
...
+  mount Quilt::Engine, at: '/path/to/react'
+end
+```
+
+### Configuration
+
+The `configure` method allows customization of the address the service will proxy to for UI rendering.
+
+```ruby
+  Quilt.configure do |config|
```
Comment (cartogram) [inline gems/quilt_rails/README.md]: Can we add a file path here?

File: gems/quilt_rails/README.md [hunk 9]
```diff
@@ -119,21 +142,204 @@ function App() {
...
+
+  def index
+    render_react
+  end
+end
+```
+
+### Engine
+
+`Quilt::Engine` provides a preconfigured controller which consumes `ReactRenderable` and provides an index route which uses it.
```
Comment (cartogram) [inline gems/quilt_rails/README.md]: Not sure if it is totally clear that mounting the Engine removes the need to use `ReactRenderable` in a controller.

File: gems/quilt_rails/README.md [hunk 10]
```diff
@@ -119,21 +142,204 @@ function App() {
...
+        └── features (optional)
+            ├── App
+            │   ├── index.js
+            │   ├── App.js
+            │   └── tests
+            │       └── App.test.js
+            │
+            ├-─ MyComponent
+            │   ├-─ index.js
+            │   ├-─ MyComponent.js
```
Comment (cartogram) [inline gems/quilt_rails/README.md]:
```suggestion
            │   ├-─ MyComponent.jsx
```

or `tsx`?

File: gems/quilt_rails/README.md [hunk 11]
```diff
@@ -119,21 +142,204 @@ function App() {
...
+            │   ├── MyComponent.scss (optional; component-scoped CSS styles, mixins, etc)
+            │   └── tests
+            │       └── MyComponent.test.js
```
Comment (cartogram) [inline gems/quilt_rails/README.md]: `jsx`

File: gems/quilt_rails/README.md [hunk 12]
```diff
@@ -119,21 +142,204 @@ function App() {
...
+
+A test would be written using Jest and `@shopify/react-testing`'s `mount` feature.
+
+```tsx
+// app/ui/components/MyComponent/tests/MyComponent.test.tsx
+import {MyComponent} from '../MyComponent';
+
+describe('MyComponent', () => {
+  it('greets the given named person', () => {
+    const wrapper = mount(<MyComponent name="Kokusho" />);
```
Comment (cartogram) [inline gems/quilt_rails/README.md]: 🐱

File: gems/quilt_rails/README.md [hunk 13]
```diff
@@ -119,21 +142,204 @@ function App() {
...
 Adds a basic `sewing-kit.config.ts` file.
 
 ## Advanced use
 
-### Interacting with the request / response in your React code
+### Testing
+
+For fast tests with consistent results, test front-end components using the tools provided by sewing-kit instead of Rails integration tests.
+
+Use [`sewing-kit test`](https://github.com/Shopify/sewing-kit/blob/master/docs/commands/test.md#L3) to run all `.test.tsx` files in the `app/ui` directory. [Jest](https://jestjs.io/) is used as a test runner, with customization available via [its sewing-kit plugin](https://github.com/Shopify/sewing-kit/blob/master/docs/plugins/jest.md).
```
Comment (cartogram) [inline gems/quilt_rails/README.md]: The flip/flopping between `.ts`/`.tsx` and `js`/`.jsx` might get confusing. Maybe we just have a consistent way of saying all of them.

File: gems/quilt_rails/README.md (hunk 3 above)
Comment (cartogram) [reply gems/quilt_rails/README.md]: We've talked about `eject` style commands in the future. I think this is sufficient until we get those out and re-write this section with those commands in mind.

**Pull request merged**